File: CnCDemo/SdkDemo/src/main/java/com/penthera/sdkdemo/notification/NotificationFactory.java
//  Copyright (c) 2013 Penthera Partners, LLC. All rights reserved.
// 
// PENTHERA CONFIDENTIAL
//
// (c) 2013 Penthera Partners Inc. All Rights Reserved.
// 
// NOTICE: This file is the property of Penthera Partners Inc.  
// The concepts contained herein are proprietary to Penthera Partners Inc.
// and may be covered by U.S. and/or foreign patents and/or patent 
// applications, and are protected by trade secret or copyright law.
// Distributing and/or reproducing this information is forbidden 
// unless prior written permission is obtained from Penthera Partners Inc.
//
package com.penthera.sdkdemo.notification;

import android.app.Notification;
import android.app.PendingIntent;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.app.NotificationCompat;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.RemoteViews;

import com.penthera.sdkdemo.MetaData;
import com.penthera.sdkdemo.R;
import com.penthera.sdkdemo.activity.SplashActivity;
import com.penthera.virtuososdk.Common;
import com.penthera.virtuososdk.Common.Notifications.DownloadStopReason;
import com.penthera.virtuososdk.client.IAsset;
import com.penthera.virtuososdk.client.IAssetManager;
import com.penthera.virtuososdk.client.IEvent;
import com.penthera.virtuososdk.client.Virtuoso;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.concurrent.atomic.AtomicReference;

public final class NotificationFactory {

	static final String LOG_TAG = NotificationFactory.class.getName();
	private static final AtomicReference<IAssetManager> _assetManager = new AtomicReference<>( null );
	private NotificationFactory() {
	}

	//types of notifications
	final static int PROGRESS_NOTIFICATION		= 0;
	final static int COMPLETED_NOTIFICATION		= 1;
	final static int STOPPED_NOTIFICATION		= 2;
	final static int PAUSED_NOTIFICATION		= 3;
	final static int RESTART_NOTIFICATION		= 4;

	/**
	 * Get the notification to be used for the intent.
	 * @param aContext the context to be used.
	 * @param aIntent the intent that was broadcast from the download service.
	 * @param aNotificationChannelID The notification channel to use.  May be null if not using notification channels.
	 * @param aAppName the applications name.
	 * @return the notification to be used
	 */
	public static Notification getNotification(Context aContext, Intent aIntent, String aNotificationChannelID, String aAppName)
	{
		if( _assetManager.get() == null){
			_assetManager.compareAndSet( null, new Virtuoso( aContext ).getAssetManager() );
		}
		Notification notification = null;
		String clientReference = null;

		try{
			ApplicationInfo ai = aContext.getPackageManager().getApplicationInfo(aContext.getPackageName(), PackageManager.GET_META_DATA);
			Bundle b =ai.metaData;

			clientReference = b.getString(Common.CLIENT_PACKAGE);

		}catch(Exception e){
			throw new RuntimeException("cannot retrieve client",e);
		}

		String action = aIntent.getAction();
		Log.d(LOG_TAG, "got action: " + action);
		if(aIntent == null || clientReference == null || action == null){
			return notification;
		}

		// If the intent action contains the NOTIFICATION_EVENT_TAG, then this is a log event broadcast sent from
		// the SDK analytics system.  For debugging, we might want to post status bar notifications, but in general,
		// these events are used to push SDK analytics events into a custom/3rd party analytics platform, and shouldn't
		// be sent to the status bar. For the purposes of this demo, we'll log them and return null, which prevents the
		// status bar notice from being shown.
		if( action.contains(Common.Notifications.NOTIFICATION_EVENT_TAG) ){

			IEvent event = aIntent.getParcelableExtra(Common.Notifications.EXTRA_NOTIFICATION_EVENT);

			Log.d(LOG_TAG,"Got event named("+event.name()+") asset("+event.assetId()+" data("+event.numericData()+")");
			return null;
		}

		// The other broadcasts are notification broadcasts specifically sent for (optional) status bar notification delivery.
		// Determine which action we are handling and create a notification for it.
		int notification_type = -1;
		IAsset file = null;

		if(action.equals(Common.START_VIRTUOSO_SERVICE)){
			notification_type = RESTART_NOTIFICATION;
		}
		else {
			boolean hasInfo = false;
			Bundle extras = aIntent.getExtras();
			int info = DownloadStopReason.NO_ERROR;

			if(extras != null) {
				if(extras.containsKey(Common.Notifications.EXTRA_NOTIFICATION_DOWNLOAD_STOP_REASON)){
					hasInfo = true;
					info = extras.getInt(Common.Notifications.EXTRA_NOTIFICATION_DOWNLOAD_STOP_REASON);
				}
				file = extras.getParcelable(Common.Notifications.EXTRA_NOTIFICATION_FILE);
			}
			final String INTENT_ACTION = action.replace(clientReference, "");

			if(INTENT_ACTION.equals(Common.Notifications.INTENT_NOTIFICATION_DOWNLOAD_COMPLETE)) {

				notification_type = COMPLETED_NOTIFICATION;
				Log.d(LOG_TAG, "DOWNLOAD COMPLETE NOTIFICATION FOR " + file.getUuid() + " stat: " + (hasInfo? info:"unknown"));
				
			} else if(INTENT_ACTION.equals(Common.Notifications.INTENT_NOTIFICATION_DOWNLOAD_START)) {

				notification_type = PROGRESS_NOTIFICATION;
				Log.d(LOG_TAG, "DOWNLOAD START NOTIFICATION FOR " + file.getUuid() + " stat: " + (hasInfo? info:"unknown"));
				
			} else if(INTENT_ACTION.equals(Common.Notifications.INTENT_NOTIFICATION_DOWNLOAD_STOPPED)) {

				if(file != null){
					Log.d(LOG_TAG, "DOWNLOAD STOP NOTIFICATION FOR " + file.getUuid() + " stat: " + (hasInfo? info:"unknown"));
				} else {
					Log.d(LOG_TAG, "DOWNLOAD STOP NOTIFICATION FOR UNKNOWN" + " stat: " + (hasInfo? info:"unknown"));
				}
				notification_type = STOPPED_NOTIFICATION;
				
			} else if(INTENT_ACTION.equals(Common.Notifications.INTENT_NOTIFICATION_DOWNLOADS_PAUSED)) {

				if(file != null){
					Log.d(LOG_TAG, "DOWNLOAD PAUSED NOTIFICATION FOR " + file.getUuid() + " stat: " + (hasInfo? info:"unknown"));
				} else {
					Log.d(LOG_TAG, "DOWNLOAD PAUSED NOTIFICATION FOR UNKNOWN" + " stat: " + (hasInfo? info:"unknown"));
				}
				notification_type = PAUSED_NOTIFICATION;
				
			} else if(INTENT_ACTION.equals(Common.Notifications.INTENT_NOTIFICATION_DOWNLOAD_UPDATE)) {

				notification_type = PROGRESS_NOTIFICATION;
				Log.d(LOG_TAG, "DOWNLOAD UPDATE NOTIFICATION FOR " + file.getUuid() + " stat: " + (hasInfo? info:"unknown"));
				
			} 
			else {
				notification_type = RESTART_NOTIFICATION;
				Log.d(LOG_TAG, "UNHANDLED NOTIFICATION ACTION "+ action);
			}
		}
		
		if(notification_type > -1){
			notification = createNotification(notification_type, aContext, aNotificationChannelID, file, aAppName);
		}
		
		return notification;
	}
	
	/**
	 * create an intent for opening the application when the user clicks on the notification.
	 * @param aContext used to get the package name
	 * @return the intent
	 */
	static Intent createIntent(Context aContext) {
		Intent intent = new Intent(aContext.getPackageName() + ".HARNESS_NOTIFICATION");
		intent.setComponent(new ComponentName(aContext.getPackageName(), SplashActivity.class.getName()));
		intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
		intent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
		intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
		intent.setFlags(Intent.FLAG_FROM_BACKGROUND);
		return intent;
	}
	
	/**
	 * populate the view which will be used by the notification
	 * @param context the context
	 * @param aTitle The title
	 * @param aQueueSize Size of the queue
	 * @param aProgress Progress of currently downloading item.
	 * @return the view
	 */
	static RemoteViews getNotificationView(Context context, String aTitle, int aQueueSize, int aProgress) {
		RemoteViews view = new RemoteViews(context.getPackageName(), R.layout.notification_layout);
		view.setTextViewText( R.id.title, aTitle );
		if(aProgress >= 0){
			view.setViewVisibility(R.id.progress, View.VISIBLE);
			view.setProgressBar(R.id.progress, 100, aProgress, false);
		}
		else {
			view.setViewVisibility(R.id.progress, View.GONE);
		}
		view.setTextViewText(R.id.queued, "Num items in queue: " + aQueueSize);
		return view;
	}
	
	/**
	 * Get the size of the queue
	 * @return size
	 */
	static int getQueueSize(){
		if(_assetManager.get() == null)
			return 0;
		return _assetManager.get().getQueue().size();
	}

	/**
	 * calculates the progress of the current download.
	 * @param aAsset the current asset downloading
	 * @return progress
	 */
	static int getDownloadProgress(final IAsset aAsset){
		if(aAsset == null)
			return -1;
		double expected = aAsset.getExpectedSize();
		if (expected <= .0001) {
			expected = aAsset.getContentLength();
		}
		return (int)  (aAsset.getCurrentSize()/expected * 100.0);
	}
	
	/**
	 * Get the assets title or an empty string
	 * @param aAsset the asset
	 * @return
	 */
	static String getAssetTitle(final IAsset aAsset){
		String title = "";

		if(aAsset != null){
			String md = aAsset.getMetadata();
			if(!TextUtils.isEmpty( md )){
				try {
					JSONObject obj = new JSONObject(md);
					title = obj.optString(MetaData.TITLE, "");
				} catch (JSONException e) {
				}
			}
		}
		return title;
	}
	
	/**
	 * Create the notification for the specified type.
	 * @param type The notification type.
	 * @param aContext The context to be used
	 * @param aNotificationChannelID The notification channel to use.  May be null if not using notification channels.
	 * @param aAsset the asset (may be null)
	 * @param aAppName name of the application - this is used in the notification title
	 * @return the notification.
	 */
	static Notification createNotification(final int type, final Context aContext, final String aNotificationChannelID,
											final IAsset aAsset, final String aAppName){
		String title = aAppName + ": ";
		int progress = -1;
		int queued = getQueueSize();

		switch(type){
		case PROGRESS_NOTIFICATION:
			progress = getDownloadProgress(aAsset);
			title += getAssetTitle(aAsset) + String.format(" ( %1$,.0f)",  aAsset.getCurrentSize());
			break;

		case COMPLETED_NOTIFICATION:
			progress = 100;
			title += getAssetTitle(aAsset) + " complete.";
			break;

		case STOPPED_NOTIFICATION:
			title += "stopped downloads.";
			break;

		case PAUSED_NOTIFICATION:
			title += "paused downloads.";
			break;

		case RESTART_NOTIFICATION:
			title += "is starting up...";
			break;
		}
		
		RemoteViews view = getNotificationView(aContext, title, queued, progress);
		Intent intent = createIntent(aContext);
		PendingIntent pIntent = PendingIntent.getActivity(aContext, 0, intent, PendingIntent.FLAG_CANCEL_CURRENT);

		Notification notification = null;
		if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && aNotificationChannelID != null && aNotificationChannelID.length() > 0) {
			notification = new Notification.Builder(aContext,aNotificationChannelID)
					.setTicker(aAppName)
					.setSmallIcon(R.drawable.ic_launcher)
					.setCustomContentView(view)
					.setContentIntent(pIntent)
					.setWhen(System.currentTimeMillis())
					.setOngoing(true).build();
		}
		else if( Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
			notification = new Notification.Builder(aContext)
					.setTicker(aAppName)
					.setSmallIcon(R.drawable.ic_launcher)
					.setCustomContentView(view)
					.setContentIntent(pIntent)
					.setWhen(System.currentTimeMillis())
					.setOngoing(true).build();
		}
		else {
			notification = new NotificationCompat.Builder(aContext)
					.setTicker(aAppName)
					.setSmallIcon(R.drawable.ic_launcher)
					.setCustomContentView(view)
					.setContentIntent(pIntent)
					.setWhen(System.currentTimeMillis())
					.setOngoing(true).build();
		}

		return notification;
	}

}

File: CnCDemo/SdkDemoKotlin/src/main/java/com/penthera/sdkdemokotlin/notification/NotificationFactory.kt
package com.penthera.sdkdemokotlin.notification

import android.app.Notification
import android.app.PendingIntent
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.util.Log
import androidx.core.app.NotificationCompat
import com.penthera.sdkdemokotlin.R
import com.penthera.sdkdemokotlin.activity.MainActivity
import com.penthera.sdkdemokotlin.catalog.ExampleMetaData
import com.penthera.virtuososdk.Common
import com.penthera.virtuososdk.client.IAsset
import com.penthera.virtuososdk.client.IAssetManager
import com.penthera.virtuososdk.client.IEvent
import com.penthera.virtuososdk.client.Virtuoso

/**
 *
 */
class NotificationFactory(private val applicationName: String) {

    companion object {
        fun channelId() = "VIRTUOSO_KOTLIN_DEMO_CHANNEL_ID"
        fun channelName() = "Demo Background Activity"
        fun channelDescription() = "Indicates activity this application will perform when the application is not open"

        private val TAG = NotificationFactory::class.java.simpleName
    }

    private var assetManager: IAssetManager? = null
    private var notificationBuilder: Notification.Builder? = null
    private var compatNotificationBuilder: NotificationCompat.Builder? = null

    private val PROGRESS_NOTIFICATION = 0
    private val COMPLETED_NOTIFICATION = 1
    private val STOPPED_NOTIFICATION = 2
    private val PAUSED_NOTIFICATION = 3
    private val RESTART_NOTIFICATION = 4
    private val FAILED_NOTIFICATION = 5

    /**
     * This defines a default intent for the app which can be used if none is provided in the request
     */
    private fun defaultNotificationIntent(context: Context) : Intent {

        val notificationIntent = Intent(context, MainActivity::class.java)
        notificationIntent.action = "foregroundservice.action.ForegroundServiceNotificationAction"
        notificationIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        return notificationIntent
    }

    fun getNotification(context: Context, intent: Intent?): Notification? {
        // A default intent is used if no intent is delivered in the request.
        val notificationIntent = intent ?: defaultNotificationIntent(context)

        var clientReference: String?

        try {
            val ai = context.packageManager.getApplicationInfo(context.packageName, PackageManager.GET_META_DATA)
            val b = ai.metaData

            clientReference = b.getString(Common.CLIENT_PACKAGE)

        } catch (e: Exception) {
            throw RuntimeException("cannot retrieve client", e)
        }

        val action = notificationIntent.action
        if (clientReference == null || action == null) {
            return null
        }

        // If the intent action contains the NOTIFICATION_EVENT_TAG, then this is a log event broadcast sent from
        // the SDK analytics system.  For debugging, we might want to post status bar notifications, but in general,
        // these events are used to push SDK analytics events into a custom/3rd party analytics platform, and shouldn't
        // be sent to the status bar. For the purposes of this demo, we'll log them and return null, which prevents the
        // status bar notice from being shown.
        if (action.contains(Common.Notifications.NOTIFICATION_EVENT_TAG)) {

            val event: IEvent? = notificationIntent.getParcelableExtra(Common.Notifications.EXTRA_NOTIFICATION_EVENT)

            event?.let {
                Log.d(TAG, "Got event named(" + it.name() + ") asset(" + it.assetId() + " data(" + it.numericData() + ")")
            }
            return null
        }

        // The other broadcasts are notification broadcasts specifically sent for (optional) status bar notification delivery.
        // Determine which action we are handling and create a notification for it.
        if (assetManager == null) synchronized(this) {
            if (assetManager == null) {
                assetManager = Virtuoso(context).assetManager
            }
        }

        var notificationType : Int
        var file: IAsset? = null

        if (action == Common.START_VIRTUOSO_SERVICE) {
            notificationType = RESTART_NOTIFICATION
        } else {
            var hasInfo = false
            val extras = notificationIntent.extras
            var info = Common.Notifications.DownloadStopReason.NO_ERROR

            if (extras != null) {
                if (extras.containsKey(Common.Notifications.EXTRA_NOTIFICATION_DOWNLOAD_STOP_REASON)) {
                    hasInfo = true
                    info = extras.getInt(Common.Notifications.EXTRA_NOTIFICATION_DOWNLOAD_STOP_REASON)
                }
                file = extras.getParcelable(Common.Notifications.EXTRA_NOTIFICATION_FILE)
            }

            when(action.replace(clientReference, "")){

                Common.Notifications.INTENT_NOTIFICATION_DOWNLOAD_COMPLETE -> {
                    notificationType = COMPLETED_NOTIFICATION
                    Log.d(TAG, "DOWNLOAD COMPLETE NOTIFICATION FOR " + file?.uuid + " stat: " + if (hasInfo) info else "unknown")

                }
                Common.Notifications.INTENT_NOTIFICATION_DOWNLOAD_START ->{
                    notificationType = PROGRESS_NOTIFICATION
                    Log.d(TAG, "DOWNLOAD START NOTIFICATION FOR " + file?.uuid + " stat: " + if (hasInfo) info else "unknown")

                }

                Common.Notifications.INTENT_NOTIFICATION_DOWNLOAD_STOPPED-> {
                    if (file != null) {
                        Log.d(TAG, "DOWNLOAD STOP NOTIFICATION FOR " + file.uuid + " stat: " + if (hasInfo) info else "unknown")
                    } else {
                        Log.d(TAG, "DOWNLOAD STOP NOTIFICATION FOR UNKNOWN" + " stat: " + if (hasInfo) info else "unknown")
                    }
                    notificationType = STOPPED_NOTIFICATION
                }

                Common.Notifications.INTENT_NOTIFICATION_DOWNLOADS_PAUSED -> {
                    if (file != null) {
                        Log.d(TAG, "DOWNLOAD PAUSED NOTIFICATION FOR " + file.uuid + " stat: " + if (hasInfo) info else "unknown")
                    } else {
                        Log.d(TAG, "DOWNLOAD PAUSED NOTIFICATION FOR UNKNOWN" + " stat: " + if (hasInfo) info else "unknown")
                    }
                    notificationType = PAUSED_NOTIFICATION
                }

                Common.Notifications.INTENT_NOTIFICATION_DOWNLOAD_UPDATE -> {
                    notificationType = PROGRESS_NOTIFICATION
                    Log.d(TAG, "DOWNLOAD UPDATE NOTIFICATION FOR " + file?.uuid + " stat: " + if (hasInfo) info else "unknown")

                }

                Common.Notifications.INTENT_NOTIFICATION_MANIFEST_PARSE_FAILED-> {
                    notificationType = FAILED_NOTIFICATION
                    Log.d(TAG, "EXCEPTIONAL CIRCUMSTANCE NOTIFICATION for asset failed to be queued while in background")
                }

                else ->{
                    notificationType = RESTART_NOTIFICATION
                    Log.d(TAG, "UNHANDLED NOTIFICATION ACTION $action")
                }
            }

        }

        var notification: Notification? = null
        if (notificationType > -1) {
            notification = createNotification(notificationType, context, file)
        }

        return notification
    }

    /**
     * Create the notification for the specified type.
     * @param type The notification type.
     * @param context The context to be used
     * @param asset the asset (may be null)
     * @return the notification.
     */
    private fun  createNotification(type: Int, context: Context, asset: IAsset?): Notification? {
        var title = "$applicationName: "
        var contentText = ""
        var progress = -1


        when (type) {
            PROGRESS_NOTIFICATION -> {
                progress = getDownloadProgress(asset)
                title += getAssetTitle(asset)
                contentText += progress.toString() + "%" + " : " + String.format(" ( %1$,.0f)", asset?.currentSize)
            }

            COMPLETED_NOTIFICATION -> {
                progress = 100
                title += getAssetTitle(asset) + " complete."
            }

            STOPPED_NOTIFICATION -> title += "stopped downloads."

            PAUSED_NOTIFICATION -> title += "paused downloads."

            RESTART_NOTIFICATION -> title += "is starting up..."

            FAILED_NOTIFICATION -> title += " asset could not be queued"
        }

        val pendingIntent = PendingIntent.getActivity(context, 0, createIntent(context), PendingIntent.FLAG_CANCEL_CURRENT)

        var notification: Notification?
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            if (notificationBuilder == null) {
                synchronized(this){
                    if (notificationBuilder == null){
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                            notificationBuilder = Notification.Builder(context, channelId())
                        } else {
                            @Suppress("DEPRECATION")
                            notificationBuilder = Notification.Builder(context)
                        }
                        notificationBuilder!!.setOnlyAlertOnce(true)
                    }
                }
            }
            var nb = notificationBuilder!!
                    .setTicker(applicationName)
                    .setSmallIcon(R.drawable.small_logo)
                    .setContentTitle(title)
                    .setContentIntent(pendingIntent)
                    .setColor(context.getColor(android.R.color.holo_blue_bright))
                    .setContentText(contentText)

            if (progress >= 0) {
                nb = nb.setProgress(100, progress, false)
            }
            notification =
                    nb.setWhen(System.currentTimeMillis())
                    .setOngoing(true).build()
        } else {
            if (compatNotificationBuilder == null) {
                synchronized(this){
                    if (compatNotificationBuilder == null) {
                        compatNotificationBuilder = NotificationCompat.Builder(context, channelId())
                        compatNotificationBuilder!!.setOnlyAlertOnce(true)
                    }
                }
            }
            var nb = compatNotificationBuilder!!
                    .setTicker(applicationName)
                    .setSmallIcon(R.drawable.small_logo)
                    .setContentTitle(title)
                    .setContentIntent(pendingIntent)
                    .setColor(context.getColor(android.R.color.holo_blue_bright))
                    .setContentText(contentText)

            if (progress >= 0) {
                nb = nb.setProgress(100, progress, false)
            }
            notification =
                    nb.setWhen(System.currentTimeMillis())
                            .setOngoing(true).build()
        }

        return notification
    }

    /**
    * calculates the progress of the current download.
    * @param asset the current asset downloading
    * @return progress
    */
    private fun getDownloadProgress(asset: IAsset?): Int {
        if (asset == null)
            return 0

        var fractionComplete = asset.fractionComplete
        fractionComplete *= 100
        if (fractionComplete > 99.0) {
            fractionComplete = 99.0
        }

        return fractionComplete.toInt()
    }

    /**
     * Get the assets title or an empty string
     * @param asset the asset
     * @return
     */
    private fun getAssetTitle(asset: IAsset?): String {

        val title = asset?.let {
            var title = ""
            val json: String = it.metadata
            if (json.isNotEmpty()) {
               title = ExampleMetaData().fromJson(json).title
            }
            title
        }

        return title ?: ""
    }

    /**
     * create an intent for opening the application when the user clicks on the notification.
     * @param aContext used to get the package name
     * @return the intent
     */
    private fun createIntent(aContext: Context): Intent {
        val intent = Intent(aContext.packageName + ".DEMO_NOTIFICATION")
        intent.component = ComponentName(aContext.packageName, MainActivity::class.java.name)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
        intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
        intent.flags = Intent.FLAG_FROM_BACKGROUND
        return intent
    }
}